map_server/src/main.cpp

```diff
@@ -67,156 +68,218 @@ class MapServer
-            mode = RAW;
-          else{
-            ROS_ERROR("Invalid mode tag \"%s\".", modeS.c_str());
-            exit(-1);
-          }
-        } catch (YAML::Exception &) {
-          ROS_DEBUG("The map does not contain a mode tag or it is invalid... assuming Trinary");
-          mode = TRINARY;
-        }
-        try {
-          doc["origin"][0] >> origin[0];
-          doc["origin"][1] >> origin[1];
-          doc["origin"][2] >> origin[2];
-        } catch (YAML::InvalidScalar &) {
-          ROS_ERROR("The map does not contain an origin tag or it is invalid.");
-          exit(-1);
-        }
-        try {
-          doc["image"] >> mapfname;
-          // TODO: make this path-handling more robust
-          if(mapfname.size() == 0)
-          {
-            ROS_ERROR("The image tag cannot be an empty string.");
-            exit(-1);
-          }
+  private:
+    ros::NodeHandle nh_;
+    ros::Publisher map_pub_;
+    ros::Publisher metadata_pub_;
+    ros::ServiceServer get_map_service_;
+    ros::ServiceServer change_map_srv_;
+    bool deprecated_;
+    std::string frame_id_;
 
-          boost::filesystem::path mapfpath(mapfname);
-          if (!mapfpath.is_absolute())
-          {
-            boost::filesystem::path dir(fname);
-            dir = dir.parent_path();
-            mapfpath = dir / mapfpath;
-            mapfname = mapfpath.string();
-          }
-        } catch (YAML::InvalidScalar &) {
-          ROS_ERROR("The map does not contain an image tag or it is invalid.");
-          exit(-1);
-        }
-      } else {
-        private_nh.param("negate", negate, 0);
-        private_nh.param("occupied_thresh", occ_th, 0.65);
-        private_nh.param("free_thresh", free_th, 0.196);
-        mapfname = fname;
-        origin[0] = origin[1] = origin[2] = 0.0;
-      }
+    /** Callback invoked when someone requests our service */
+    bool mapCallback(nav_msgs::GetMap::Request  &req,
+                     nav_msgs::GetMap::Response &res )
+    {
+      // request is empty; we ignore it
+
+      // = operator is overloaded to make deep copy (tricky!)
+      res = map_resp_;
+      ROS_INFO("Sending map");
+
+      return true;
+    }
 
-      ROS_INFO("Loading map from image \"%s\"", mapfname.c_str());
-      try
+    /** Callback invoked when someone requests to change the map */
+    bool changeMapCallback(nav_msgs::LoadMap::Request  &request,
+                           nav_msgs::LoadMap::Response &response )
+    {
+      if (loadMapFromYaml(request.map_url))
       {
-          map_server::loadMapFromFile(&map_resp_,mapfname.c_str(),res,negate,occ_th,free_th, origin, mode);
+        response.result = response.RESULT_SUCCESS;
       }
-      catch (std::runtime_error e)
+      else
       {
-          ROS_ERROR("%s", e.what());
-          exit(-1);
+        response.result = response.RESULT_UNDEFINED_FAILURE;
       }
+      ROS_INFO("Changed map to %s", request.map_url.c_str());
```

Comment:
Seems odd to tell the user we changed the map if the loadMapFromYaml failed? Should this message be moved into the if() statement only? (it looks like we already get error messages from the loadMapFromYaml function).